File: code/lazyEvaluation.js
let Lazy = require('lazy.js');

function square(x) { return x * x; }
function isEven(x) { return x % 2 === 0; }

let array = Lazy.range(100).toArray();
//console.log(array);
let result = Lazy(array).map(square).filter(isEven).first(5);
//console.log(result);

result.each((e) => {
    console.log(e);
});
console.log(result.toArray());

let fibonacci = Lazy.generate(function() {
    var x = 1,
        y = 1;
    return function() {
      var prev = x;
      x = y;
      y += prev;
      return prev;
    };
  }());
console.log(fibonacci.take(5).toArray());
  

File: code/lazyEvaluation.py
def fibonacci(n): # 生成器函数 - 斐波那契
    a, b, counter = 1, 1, 0
    while True:
        if (counter > n): 
            return
        yield a
        a, b = b, a + b
        counter += 1
f = fibonacci(10) # f 是一个迭代器，由生成器返回生成

for i in f:
    print(i, end=" ")  # 1 1 2 3 5 8 13 21 34 55 89